File: backend-ml/multimodelHelper.py
# import speech_recognition as sr
import google.generativeai as genai
from dotenv import load_dotenv
import os
from gtts import gTTS
import PIL.Image
import time
from pydub import AudioSegment
import tempfile

# Load environment variables
load_dotenv()

# Set the Google API Key
GOOGLE_API_KEY = os.getenv('GOOGLE_API_KEY')

# Function to handle voice input using a file
# def voice_input(audio_path):
#     try:
#         # Convert WebM to WAV
#         audio = AudioSegment.from_file(audio_path, format="webm")
#         with tempfile.NamedTemporaryFile(delete=False, suffix=".wav") as temp_wav_file:
#             wav_path = temp_wav_file.name
#             audio.export(wav_path, format="wav")

#         print(f"WAV file created at: {wav_path}")

#         r = sr.Recognizer()
#         with sr.AudioFile(wav_path) as source:
#             audio_data = r.record(source)
#         try:
#             text = r.recognize_google(audio_data)
#             print("Transcribed: ", text)
#             return text
#         except sr.UnknownValueError:
#             print("Sorry, could not understand the audio")
#             return "Could not understand the audio"
#         except sr.RequestError as e:
#             print(f"Error with Google Speech Recognition service: {e}")
#             return "Error with speech recognition service"
#     except Exception as e:
#         print(f"Error processing audio: {str(e)}")
#         return f"Error processing audio: {str(e)}"
#     finally:
#         # Clean up temporary files
#         try:
#             os.unlink(wav_path)
#         except:
#             pass

# Function to generate response from LLM model for voice chat
def llm_model_audio(user_text):
    genai.configure(api_key=GOOGLE_API_KEY)

    model = genai.GenerativeModel(model_name='gemini-1.5-pro')
    prompt1 = """
    You are a helpful assistant named 'ShauryaNova' developed by Ayush Shaurya Jha. 
    You are supposed to answer accurately and precisely to the user's question. 
    Now, the user query begins:
    """
    content = f"{prompt1}\n{user_text}"
    response = model.generate_content(content)
    result = response.text
    result_cleaned = result.replace('*', '')
    return result_cleaned

# Function to generate response from LLM model for image input
def llm_model_image(user_text, path):
    genai.configure(api_key=GOOGLE_API_KEY)
    model = genai.GenerativeModel(model_name='gemini-1.5-pro')
    prompt1 = """
    You are a helpful assistant named 'ShauryaNova' developed by Ayush Shaurya Jha. 
    You are supposed to answer accurately and precisely to the user's question 
    and image. Now, the user query begins:
    """
    sample_file = PIL.Image.open(path)
    prompt = f"{prompt1}\n{user_text}"
    response = model.generate_content([prompt, sample_file])
    result = response.text
    return result

# Function to generate response from LLM model for video input
def llm_model_video(user_text, video_path):
    genai.configure(api_key=GOOGLE_API_KEY)
    model = genai.GenerativeModel(model_name='gemini-1.5-pro')

    prompt1 = """
    You are a helpful assistant named 'ShauryaNova' developed by Ayush Shaurya Jha.
    You are supposed to answer accurately and precisely to the user's question 
    and video. Now, the user query begins:
    """

    if not video_path:
        raise ValueError("Video path cannot be None")

    print(f"Uploading file: {video_path}")
    video_file = genai.upload_file(path=video_path)
    print(f"Completed upload: {video_file.uri}")

    # Check if file processing is done
    while video_file.state.name == "PROCESSING":
        print("Processing video...")
        time.sleep(10)
        video_file = genai.get_file(video_file.name)

    if video_file.state.name == "FAILED":
        raise ValueError("Video processing failed")

    prompt = f"{prompt1}\n{user_text}" if user_text else prompt1
    response = model.generate_content([video_file, prompt], request_options={"timeout": 600})
    result = response.text
    return result

# Function to convert text response to speech
def text_to_speech(text):
    tts = gTTS(text=text, lang="en")
    tts.save("speech.mp3")

File: backend/routes/multimodelHelper.js
const axios = require('axios');
const fs = require('fs').promises;
const FormData = require('form-data');
const { GoogleGenerativeAI } = require('@google/generative-ai');
const dotenv = require('dotenv');
const sharp = require('sharp');
const path = require('path');
const os = require('os');
const gTTS = require('gtts');
const ffmpeg = require('fluent-ffmpeg');
const ffmpegPath = require('ffmpeg-static');
const ffprobePath = require('@ffprobe-installer/ffprobe').path;

dotenv.config();

// Set FFmpeg and FFprobe paths
ffmpeg.setFfmpegPath(ffmpegPath);
ffmpeg.setFfprobePath(ffprobePath);

const GOOGLE_API_KEY = process.env.GOOGLE_API_KEY;
const genAI = new GoogleGenerativeAI(GOOGLE_API_KEY);

async function llmModelAudio(userText) {
  const model = genAI.getGenerativeModel({ model: "gemini-1.5-pro" });
  const prompt = `
    You are a helpful assistant named 'ShauryaNova' developed by Ayush Shaurya Jha. 
    You are supposed to answer accurately and precisely to the user's question. 
    Now, the user query begins:
    ${userText}
  `;

  const result = await model.generateContent(prompt);
  const response = await result.response;
  return response.text().replace(/\*/g, '');
}

async function llmModelImage(userText, imagePath) {
  const model = genAI.getGenerativeModel({ model: "gemini-1.5-pro" });
  const prompt = `
    You are a helpful assistant named 'ShauryaNova' developed by Ayush Shaurya Jha. 
    You are supposed to answer accurately and precisely to the user's question 
    and image. Now, the user query begins:
    ${userText}
  `;

  const imageData = await sharp(imagePath).toBuffer();
  const imageParts = [
    {
      inlineData: {
        data: imageData.toString('base64'),
        mimeType: 'image/jpeg',
      },
    },
  ];

  const result = await model.generateContent([prompt, ...imageParts]);
  const response = await result.response;
  return response.text();
}

async function llmModelVideo(userText, videoPath) {
  const model = genAI.getGenerativeModel({ model: "gemini-1.5-pro" });
  const prompt = `
    You are a helpful assistant named 'ShauryaNova' developed by Ayush Shaurya Jha.
    You are supposed to answer accurately and precisely to the user's question 
    and video. Now, the user query begins:
    ${userText}
  `;

  if (!videoPath) {
    throw new Error("Video path cannot be None");
  }

  console.log(`Processing video: ${videoPath}`);

  // Extract frames from the video
  const frameCount = 5; // Number of frames to extract
  const frames = await extractFrames(videoPath, frameCount);

  // Process frames and create image parts for the model
  const imageParts = await Promise.all(frames.map(async (framePath) => {
    const imageData = await sharp(framePath).resize(512, 512).toBuffer();
    return {
      inlineData: {
        data: imageData.toString('base64'),
        mimeType: 'image/jpeg',
      },
    };
  }));

  // Generate content using the model
  const result = await model.generateContent([prompt, ...imageParts]);
  const response = await result.response;

  // Clean up temporary frame files
  await Promise.all(frames.map(framePath => fs.unlink(framePath)));

  return response.text();
}

// Helper function to extract frames from a video
function extractFrames(videoPath, frameCount) {
  return new Promise((resolve, reject) => {
    const tempDir = os.tmpdir();
    const frames = [];

    ffmpeg(videoPath)
      .on('filenames', function(filenames) {
        frames.push(...filenames.map(filename => path.join(tempDir, filename)));
      })
      .on('end', function() {
        resolve(frames);
      })
      .on('error', function(err) {
        console.error('Error:', err);
        reject(err);
      })
      .screenshots({
        count: frameCount,
        folder: tempDir,
        filename: 'frame-%i.png'
      });
  });
}

async function textToSpeech(text) {
  return new Promise((resolve, reject) => {
    try {
      console.log("textToSpeech function called", text);
      
      const outputFile = path.join(__dirname, '..', 'speech.mp3');
      const gtts = new gTTS(text, 'en');
      
      gtts.save(outputFile, (err) => {
        if (err) {
          console.log("Error in textToSpeech function", err);
          reject(err);
        } else {
          console.log('Audio content written to file:', outputFile);
          resolve();
        }
      });
    } catch (error) {
      console.log("Error in textToSpeech function", error);
      reject(error);
    }
  });
}

module.exports = {
  llmModelAudio,
  llmModelImage,
  llmModelVideo,
  textToSpeech
};
